### PR TITLE
Use single codepath for sending transactions to a local and remote nodes

### LINKED ISSUE #529: Use a single flow to send a transaction for remote and local nodes
# Problem

It is described in [this comment](https://github.com/status-im/status-go/issues/338#issuecomment-331389930). 

The problem is that having two separate flows to send a transaction introduces subtle differences that may be hard to debug, maintain and can lead to inconsistencies.

# Implementation

Create a single flow in status-go to send all transactions via `eth_sendRawTransaction` regardless which node is used.

# Acceptance Criteria

1. There is a single flow to send a transaction,
1. The whole code related to gas estimation, getting a nonce etc. is in one place,
1. Sending a transaction works with local and remote nodes.

### PR DESCRIPTION
**Pull request by dshulyak**
Fixes #529

Previously procedure for sending transactions was different based on
which type of node was used, local (light node) or remote.

This change removes separate code path for working with light node,
and the only difference will be backend for rpc client. In case of
light node all of the communication will be done over inproc dial.

Also there is a couple of related changes in this PR:
- new EthereumTransactor that provides higher level API for working
  with ethereum network, and it is fully conformant with ethclient
- new test rpc service that improves flexibility and coverage of
  txqueue manager tests

Closes: https://github.com/status-im/status-go/issues/529

### CODE REVIEW

**Comment by dshulyak** [comment]
@adambabik hello, i believe this change is ready for initial review. 
I plan to add more tests for different edge cases and do other refactoring that is described in https://github.com/status-im/status-go/issues/338 , but i think it would be better to make separate PRs for those things.

**Comment by dshulyak** [comment]
Looks like i asked prematurely, e2e tests are failing with relevant error.

**Comment by dshulyak** [comment]
At the moment only one test fails - TestCompleteMultipleQueuedTransactions, and I am pretty sure that I know the root cause of this issue. LesBackend internally uses txPool.GetNonce method when it sets the nonce for an incoming transaction. This method looks up a recent nonce at the local map of the pool - pool.nonce. But this method is not exported, and I have to use GetTransactionCount that is not consistent with txPool.GetNonce. I tried to verify it on develop branch:

| tx | txPool.GetNonce | GetTransactionCount(pending) |
|---|------------------|---------------------------------|
|  1 |               0 |                            0 |
|  2 |               1 |                            0 |
|  3 |               2 |                            0 |

I think I can add another interceptor, to a status rpc.Client, for GetTransactionCount method which will use LesBackend and GetPoolNonce. But I am pretty sure that it is a bug in LightEthereum client.

https://github.com/ethereum/go-ethereum/issues/2880

**Comment by adambabik** [comment]
Hey @dshulyak. Thank you for your contribution. I skimmed through the PR and it looks great! I will submit a detailed review tomorrow.

Regarding the nonce, in the current develop, the code that sends a transaction when using a remote node uses the same JSON-RPC method and arguments to get the nonce as before so the tests should pass just fine, unless these tests use a local node... In this case, a pool is used: https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L1096-L1102 and this go-ethereum issue seems like a root cause. You can hot patch code in `vendor/` and see if it fixes this problem. We can incorporate this patch if it will turn out to be necessary.

**Comment by dshulyak** [comment]
@adambabik i don't see a reason why it won't be possible to reproduce this bug with upstream node.  maybe it will be required to increase the number of concurrent transactions

**Comment by adambabik** [comment]
@dshulyak you're right, this bug should be also visible when using a remote node. We may not have an automated test for the remote node, though. 

To reviewers: the tests were failing because they use a local node and, before changes, they were using LES backend directly instead of the public API (that is `eth_sendRawTransaction` method). It turns out that the public API has a bug and does not use the transaction pool to return a correct nonce value.

File: geth/txqueue/ethtxclient.go
Comment (adambabik) [inline geth/txqueue/ethtxclient.go]: To keep it consistent, maybe we can call it `EthTransactor`, similarly to `EthTxClient`?

File: geth/txqueue/txqueue_manager.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager.go]: Maybe `defaultTimeout` to match `defaultGas` naming convention?

File: geth/txqueue/txqueue_manager.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager.go]: This is fine but `else` can be omitted:

```go
gasPrice := (*big.Int)(args.GasPrice)
if gasPrice == nil {
  // ...
}
```

**Comment by dshulyak** [comment]
I think that it is possible yo retry tx on "known transaction" error, both light and full nodes return it. But, i need to understand if there are any other side effects of doing it.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: I think it should be possible to get rid of this map and only assert `errCompleteTransaction` to be always `nil`.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: As this block repeats three times, maybe we can extract it to a separate `TxQueueTestSuite` method?

File: geth/txqueue/fake/txservice.go
Comment (adambabik) [inline geth/txqueue/fake/txservice.go]: Maybe we can call it `FakePublicTransactionPoolAPI` to make it clearer what it is?

File: geth/txqueue/fake/txservice.go
Comment (adambabik) [inline geth/txqueue/fake/txservice.go]: Is `big` required?

File: geth/txqueue/fake/txservice.go
Comment (adambabik) [inline geth/txqueue/fake/txservice.go]: I believe `gomock` is not required before the package name here.

File: geth/txqueue/fake/mock.go
Comment (adambabik) [inline geth/txqueue/fake/mock.go:1]: Please add a proper entry to `mock` target: https://github.com/status-im/status-go/blob/develop/Makefile#L101-L105

**Review by adambabik — CHANGES_REQUESTED** [review]
It looks great! I added a bunch of comments but they are minor suggestions or code style.

I will create a separate issue for this effort as it's related to #338 indeed, but not exactly.

**Comment by adambabik** [comment]
@dshulyak I created a separate issue for this PR: https://github.com/status-im/status-go/issues/529.

**Comment by dshulyak** [comment]
@adambabik thanks for a review, i pushed an update

**Comment by andytudhope** [comment]
Hey @dshulyak! Adam suggested that we offer a bounty for this issue because you had done such good work. I have created and funded it and linked this PR to the issue with a minor edit. Could you please log into http://openbounty.status.im/ click on your profile picture, select to `Update Your Payment Details` and make sure to click the `Update` button after putting in an address you control?

**Comment by dshulyak** [comment]
that's awesome @andytudhope, I updated my address as you suggested

**Comment by dshulyak** [comment]
Can someone re-trigger a job on travis? It failed due to `'https://github.com/alecthomas/gometalinter/': Failed to connect to github.com port 443: Connection timed out`

**Comment by divan** [comment]
@dshulyak done.

File: geth/txqueue/txqueue_manager.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager.go]: Minor: I noticed that maybe this arguments grouping could be better: `m.completeTransaction(queuedTx, selectedAccount, password)`

File: geth/txqueue/txqueue_manager.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager.go]: I would move this log just above `m.accountManager.VerifyAccountPassword` or remove at all.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: After these changes, we do not expect any errors so we can verify if the returned hash is equal to `tx.Hash`, right?

Comment (adambabik) [reply geth/txqueue/txqueue_manager_test.go]: Also, a few lines above, there is `txQueueManager.SetTransactionReturnHandler()`. I think we should also change it to expect `err` to be `nil`.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: Shouldn't it be:
```go
s.Equal(txCount-1, inprogressTx, "txs expected to be reported as inprogress")
```
?

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: This test does not test a real scenario. `keystore.ErrDecrypt` is returned by `SendRawTransaction` while it should be by `VerifyAccountPassword`. I would recommend to change that and also make sure that `SendRawTransaction` is not called.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go]: Minor: as there are multiple `var`s, it's beneficial to group them with `var ()`.

File: vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go
Comment (adambabik) [inline vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:979]: We need to add this to go-ethereum patches. cc @divan

**Review by adambabik — CHANGES_REQUESTED** [review]
Sorry about doing this review in two iterations but I found a bit more time just today.

Minor remarks overall, just please make sure to look at the comments relating to `TestInvalidPassword` and `TestCompleteTransaction`.

File: vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go
Comment (dshulyak) [reply vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:979]: I think there is a way to get rid of it. we can maintain nonce locally for each address that status is using. increment it on every successful transaction, and reset in case if it was changed remotely. it is easy with new addr locker and will prevent issues both with upstream and local nodes.

will do a separate pr after additional testing

Comment (adambabik) [reply vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:979]: It may introduce a lot of complexity into status-go and something that potentially can break in the future if go-ethereum implementation changes. But if you find some elegant solution that would be awesome.

I have seen this PR https://github.com/ethereum/go-ethereum/pull/15794 and read the whole discussion in https://github.com/ethereum/go-ethereum/issues/2880 and it looks like a separate JSON-RPC method is the best solution that can preserve backward compatibility of `eth_getTransactionCount` and allows to get the number of pending transactions from the pool.

Comment (dshulyak) [reply vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:979]: adding a separate method also seemed like a good idea to me, but it may take some time until it will be merged and a new geth release will be available. and even then status app needs to be always aware that upstream node is not an old geth release or another client, e.g parity.

File: testing/testing.go
Comment (adambabik) [inline testing/testing.go:220]:
```go
case <-time.After(d):
  return ErrTimeout
```
will do the same.

File: geth/txqueue/txqueue_manager_test.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager_test.go:310]: I am not a fan of one char variables. In such contexts, `done := make(chan struct{})` works pretty well and indicates an intention clearly.

**Review by adambabik — APPROVED** [review]
LGTM 👏

**Comment by dshulyak** [comment]
thanks, I will change those things in one of the pr's that is based on this one.

travis failed because of https://github.com/status-im/status-go/issues/539

File: vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go
Comment (divan) [reply vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:979]: Yeah, that should be implemented via patches to keep track of our changes into the `go-ethereum`. This is blocked by https://github.com/status-im/status-go/pull/492, which will be merged tomorrow (internally agreed deadline).

**Review by divan — APPROVED** [review]
Great work!
The modification to the vendored `go-ethereum` should be made through patches, so this PR is currently blocked by https://github.com/status-im/status-go/pull/492. Will try to resolve this block asap.

File: geth/txqueue/txqueue_manager.go
Comment (adambabik) [inline geth/txqueue/txqueue_manager.go]: I think it's good. The other way is to pass existing `gas` value to `EstimateGas` arguments and it will take it as upper bound, however, a lower value might be returned.

Skipping `EstimateGas` call if gas is defined should be better, though.

Comment (dshulyak) [reply geth/txqueue/txqueue_manager.go]: also the idea was to keep EthTxClient compatible with https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L36 , it is not of high priority of course

**Comment by divan** [comment]
As https://github.com/status-im/status-go/pull/492 has been merged, changes to `internal/ethapi/api.go` should be implemented as a patch (see https://github.com/status-im/status-go/tree/develop/geth-patches directory). @dshulyak when you have a time, could you make this change?

The exact way of applying patches to our fork might be changed in https://github.com/status-im/status-go/pull/551, but so far patch files are in `geth-patches` dir.

**Comment by dshulyak** [comment]
@divan pushed the patch

**Comment by divan** [comment]
@dshulyak perfect, thanks! Actually, I think we should not wait for https://github.com/status-im/status-go/pull/551, because the process of applying patches should be repeatable anyway, so we can add new patches meanwhile. Thus, unblocking this PR now.

**Comment by divan** [comment]
Trying to understand why tests on Travis time outs before merging.

**Comment by divan** [comment]
@dshulyak awesome job.

**Pull request merged**
